File: src/main/java/challenge/easy/restoreFinishingOrder/RestoreFinishingOrder.py
class Solution:
    def recoverOrder(self, order: List[int], friends: List[int]) -> List[int]:
        sorted_list = []

        for i in order:
            j = self._binary_search(friends, i)
            if j != -1 and i == friends[j]:
                sorted_list.append(i)

        return sorted_list

    def _binary_search(self, ints: List[int], value: int) -> int:
        l = 0
        r = len(ints) - 1

        while l <= r:
            m = l + (r - l) // 2

            if ints[m] == value:
                return m

            if ints[m] < value:
                l = m + 1
            else:
                r = m - 1

        return -1

File: src/main/java/challenge/easy/restoreFinishingOrder/RestoreFinishingOrder.c
#include <stdio.h>
#include <stdlib.h>

int compareInts(const void* a, const void* b)
{
    int ia = *(const int*)a;
    int ib = *(const int*)b;
    return (ia > ib) - (ia < ib);
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* recoverOrder(int* order, int orderSize, int* friends, int friendsSize, int* returnSize)
{
    int* sorted = (int*)malloc(orderSize * sizeof(int));
    if (sorted == NULL)
    {
        *returnSize = 0;
        return NULL;
    }

    int index = 0;

    for (int i = 0; i < orderSize; i++)
    {
        int* found = (int*)bsearch(&order[i], friends, friendsSize, sizeof(int), compareInts);
        if (found != NULL)
        {
            sorted[index++] = order[i];
        }
    }

    if (index == 0)
    {
        free(sorted);
        *returnSize = 0;
        return NULL;
    }

    int* resized = (int*)realloc(sorted, index * sizeof(int));
    if (resized != NULL)
    {
        sorted = resized;
    }

    *returnSize = index;
    return sorted;
}